---
title: "The smallest intron"
author: "Anqi Dai"
date: "3/25/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
library(GenomicFeatures)
library(rtracklayer)
library(GenomicRanges)
```

```{r}
gtf <- makeTxDbFromGFF("../../../../../reference/v27/gencode.v27.annotation.gtf") #change me!
exons <- exonsBy(gtf, by="gene")

#make introns
exons <- GenomicRanges::reduce(exons)
exons <- exons[sapply(exons, length) > 1]

introns <- lapply(exons, function(x) {
    #Make a "gene" GRange object
    gr = GRanges(seqnames=seqnames(x)[1], ranges=IRanges(start=min(start(x)),
        end=max(end(x))), 
        strand=strand(x)[1])
    db = disjoin(c(x, gr))
    ints = db[countOverlaps(db, x) == 0]
  
    ints
})
introns <- GRangesList(introns)
```

```{r}
# extract the intron length and the gene ID out from the grange object and make that a df
res <- lapply(names(introns), function(name){
  ret = data_frame(
    ensembl_gene_id = name,
    intronLen = width(introns[[name]])
  ) %>%
    arrange(intronLen)
}) %>%
  bind_rows

# clean the ID and keep only the two shortest intron length
short <- res %>%
  mutate(ensembl_gene_id = str_replace_all(string = ensembl_gene_id, pattern = '\\..*$', replacement = '')) %>%
  group_by(ensembl_gene_id) %>%
  summarise( intronLen = min(intronLen))
```

```{r}
# load the sig RI locus in day 3 and check out the length of that RI
d3RI <- read_csv('../output/all_tools_sig05_locus_info_day3_with_all_PSI_values.csv') %>%
  filter(Psi_14 > 0.1 & Psi_89 > 0.1  & abs(deltaPSI) > 0.1) %>%
  dplyr::select(locus, type, external_gene_name) %>%
  filter(type == 'RI') %>%
  separate(locus, into = c('chr','coord'), sep = ':') %>%
  separate(coord, into = c('start','end'),sep = '-', convert = T) %>%
  mutate(RIlen = end - start)


# collapse the df to see the genes that have more than one RI event
unique_gene <- d3RI %>%
  group_by(external_gene_name) %>%
  arrange(RIlen, .by_group = TRUE) %>%
  summarise(RI_len_all = paste0(RIlen, collapse = ","),
            cnt = n()) %>%
  arrange(desc(cnt))

# get the ensemble gene id for those sig RI locus
library(biomaRt)

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# the ensembl_gene_id of the genes that you are interested in (unique)
genes <- d3RI %>%
  distinct(external_gene_name) %>%
  pull(external_gene_name)

# retrieve the feature information of the listed attributes
symbol <- getBM(filters = "external_gene_name",
                attributes = c('ensembl_gene_id','external_gene_name'),
                values = genes, 
                mart = mart)

# the unique ensemble gene ids
symbol <- symbol %>%
  distinct(external_gene_name, .keep_all = T)
  
```

```{r}
sig_RI <- unique_gene %>%
  left_join(symbol, by = 'external_gene_name') %>%
  left_join(short, by = 'ensembl_gene_id') %>%
  dplyr::select(external_gene_name, shortestIntron = intronLen, RI_len_all, cnt, ensembl_gene_id)
```

```{r}
# what is happening with MEX3A
sig <- d3RI %>%
  filter(RIlen == 112 & external_gene_name == 'MEX3A')
```

