---
title: "Pooled KD 1 pair DE"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
library(tidyverse)
library(ggpubr)
library(kableExtra)
library(knitr)
library(biomaRt)
library(Biobase)
```

Doing 1 pair DE pooling KD together to contrast with control.

## One pair DE

Pooling 14 and 89 together to contrast with control.

```{r}
# remake the pheno table with KD1 and KD2 together as KD
pheno <- pData(read_rds('../../08_findingGenes/data/Norm_salmon_cts_replicate1_eset_unfiltered.RDS')) %>%
  rownames_to_column %>%
  mutate(Status = str_replace(Status, 'KD1|KD2', 'KD')) %>%
  column_to_rownames('rowname')

write.csv(pheno, '../data/pheno_pooled_KD.csv', row.names = T,  quote = F)
```

```{bash}
# run the DE using the previously filtered data
#detk-de deseq2 "counts ~ Status" ../../10_norval_annotation_quantification/data/normal_quant_salmon_raw_cts_filtered.csv ../data/pheno_pooled_KD.csv  > ../output/03_pooledKD_DE.csv

```

## DE results

Selecting padj < 0.05 as the significant genes.

```{r}
# look at the DE results
pooled <- read_tsv('../output/03_pooledKD_DE.csv')

sig <- pooled %>%
    filter(Status__KD__padj < 0.05) %>%
    pull(X)

# check how these genes look in volcano plot in 14 data
pooled_plot <- pooled %>%
  filter(abs(Status__KD__log2FoldChange ) < 5) %>%
  mutate(Significance = factor(ifelse(X %in% sig, 'sig', 'no' ))) %>%
  mutate(log10Padj = -log10(Status__KD__padj))

# plot the volcano plot
pooled_plot %>%
  ggscatter(x = 'Status__KD__log2FoldChange',
            y = 'log10Padj',
            color = 'Significance',
            alpha = 0.1,
            xlab =  'log2 Fold Change',
            ylab = '-log10(PADJ)',
            title = 'Volcano plot for pooled KD VS control'
            ) + 
  geom_vline(xintercept = 1, linetype = 'dashed') + 
  geom_vline(xintercept = -1, linetype = 'dashed') +
  scale_color_manual(values = c('#00468B', '#EC0000')) 
```

## Filter on the sig genes

* padj < 0.05
* 1 < abs(log2FC) < 5

```{r}
# again filter on the abs fold change
Sig <- pooled %>%
    filter(Status__KD__padj < 0.05) %>%
    filter(abs(Status__KD__log2FoldChange ) < 5) %>%
    filter(abs(Status__KD__log2FoldChange ) > 1) %>%
    pull(X)

write.table(Sig, '../output/03_pooledKD_sig_genes_IDs.txt', quote = F, col.names = F, row.names = F)


##  the genes that are up in KD
up <- pooled %>%
    filter(Status__KD__padj < 0.05) %>%
    filter(abs(Status__KD__log2FoldChange ) < 5) %>%
    filter(abs(Status__KD__log2FoldChange ) > 1) %>%
    filter(Status__KD__log2FoldChange > 0) %>%
    pull(X)

write.table(up, '../output/03_pooledKD_sig_genes_IDs_UP.txt', quote = F, col.names = F, row.names = F)


##  the genes that are down in KD
down <- pooled %>%
    filter(Status__KD__padj < 0.05) %>%
    filter(abs(Status__KD__log2FoldChange ) < 5) %>%
    filter(abs(Status__KD__log2FoldChange ) > 1) %>%
    filter(Status__KD__log2FoldChange < 0) %>%
    pull(X)

write.table(down, '../output/03_pooledKD_sig_genes_IDs_DOWN.txt', quote = F, col.names = F, row.names = F)
```

## Metascape result

### All genes 

```{r out.width=800}
knitr::include_graphics('../output/Metascape/03_pooledKD/Enrichment_heatmap/HeatmapSelectedGO.png')
```

### Genes upregulated in KD

```{r out.width=800}
knitr::include_graphics('../output/Metascape/03_pooledKD_up/all/Enrichment_heatmap/HeatmapSelectedGO.png')
```

*SIRT1 negatively regulates rRNA expression*

Expression of rRNA genes is coupled to the overall metabolism of the cell by the NAD-dependent histone deacetylase SIRT1, a component of the Energy-dependent Nucleolar Silencing Complex (eNoSC) (Murayama et al. 2008, reviewed in Salminen and Kaarniranta 2009, Grummt and Voit 2010). eNoSC comprises Nucleomethylin (NML), SIRT1, and the histone methylase SUV39H1 (Murayama et al. 2008). Deacetylation and methylation of histone H3 in the chromatin of a rRNA gene by eNoSC causes reduced expression of the gene. When glucose is low, NAD is high (NADH is low), activity of SIRT1 is high, and activity of rRNA genes is reduced. It is hypothesized that eNoSC forms on a nucleosome containing dimethylated lysine-9 on histone H3 (H3K9me2) and then eNoSC deacetylates and dimethylates the adjacent nucleosome, thus catalyzing spreading of H3K9me2 throughout the gene.

*The hits genes of this pathway*

```{r}
up_GO <- read_csv('../output/Metascape/03_pooledKD_up/all/Enrichment_GO/GO_AllLists.csv')

sirti <- up_GO %>%
  filter(Description == 'SIRT1 negatively regulates rRNA expression') %>%
  pull(Hits) %>%
  strsplit( '\\|') %>% # you have to escape the | it's a special characer in regex
  unlist

# get the gene information

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

symbol <- getBM(filters = "external_gene_name",
                attributes = c('ensembl_gene_id',
                               'external_gene_name',
                               'chromosome_name',
                               'start_position',
                               'end_position', 
                               'description'),
                values = sirti, 
                mart = mart)
symbol <- symbol %>%
  rename(chr = chromosome_name) %>%
  filter(! grepl('CHR', chr))

symbol %>%
  kable() %>%
  kable_styling()
```

### Genes downregulated in KD

```{r out.width=800}
knitr::include_graphics('../output/Metascape/03_pooledKD_down/all/Enrichment_heatmap/HeatmapSelectedGO.png')
```

*small molecule biosynthetic process*

The chemical reactions and pathways resulting in the formation of small molecules, any low molecular weight, monomeric, non-encoded molecule. Small molecules in GO include monosaccharides but exclude disaccharides and polysaccharides.


## Retrieve the sig pathways that have TP53

### All genes

```{r}
# positive regulation of cell death
# check what those genes are
all_GO <- read_csv('../output/Metascape/03_pooledKD/Enrichment_GO/GO_AllLists.csv')

prcd <- all_GO %>%
  filter(Description == 'positive regulation of cell death') %>%
  pull(Hits) %>%
  strsplit( '\\|') %>% # you have to escape the | it's a special characer in regex
  unlist

# get the gene information

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

symbol <- getBM(filters = "external_gene_name",
                attributes = c('ensembl_gene_id',
                               'external_gene_name',
                               'chromosome_name',
                               'start_position',
                               'end_position', 
                               'description'),
                values = prcd, 
                mart = mart)
symbol <- symbol %>%
  rename(chr = chromosome_name) %>%
  filter(! grepl('CHR', chr))
```

*How many*

```{r}
# get all the sig enriched pathways that have TP53 in it
target <- all_GO %>%
  filter(Enrichment > 1.3) %>%
  filter(grepl('TP53', x = Hits))  

nrow(target)

target %>%
  dplyr::select(-GeneID) %>%
  kable() %>%
  kable_styling()
```

### Up genes

*How many*

```{r}
up_GO %>%
  filter(Enrichment > 1.3) %>%
  filter(grepl('TP53', x = Hits)) %>%
  nrow
```

```{r}
up_GO %>%
  filter(Enrichment > 1.3) %>%
  filter(grepl('TP53', x = Hits)) %>%
  dplyr::select(-GeneID) %>%
  kable() %>%
  kable_styling()
```


### Down genes

```{r}
down_GO <- read_csv('../output/Metascape/03_pooledKD_down/all/Enrichment_GO/GO_AllLists.csv')
```

*How many*

```{r}
down_GO %>%
  filter(Enrichment > 1.3) %>%
  filter(grepl('TP53', x = Hits)) %>%
  nrow
```

```{r}
down_GO %>%
  filter(Enrichment > 1.3) %>%
  filter(grepl('TP53', x = Hits)) %>%
  dplyr::select(-GeneID) %>%
  kable() %>%
  kable_styling()
```