---
title: "Retrieve unique backgroud introns"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
library(tidyverse)
library(knitr)
library(kableExtra)
```

## Extract introns from GTF file

```{r}
# Since there seems to be some bugs in pybedtools that it's not yielding reliable result. We (Adam) decided to quit using that all together. He wrote the script 000_ and we will use the data generated from that.
#./000_extract_introns_from_gtf.py  ../../../../../reference/v27/gencode.v27.annotation.gtf > ../data/introns.bed
```

```{r}
# format the result from running that python script.
# form the locus and the intron len column.
all_introns <- read_tsv('../data/introns.bed', 
                        col_names = c('chr',
                                      'start',
                                      'end',
                                      'strand',
                                      'ensembl_gene_id'))  %>%
  mutate(ensembl_gene_id = str_replace_all(string = ensembl_gene_id, pattern = '\\..*$', replacement = ''),
         locus = paste(chr,paste(start, end, sep = '-'), sep = ':' ),
         intronLen = end - start) 
```

## Select genes to be background ones

```{r}
# selecting genes to be the background ones
# only removing the rows that have all zeros

# detk-filter  -o raw_salmon_cts_day3_matrix_nonzero_all_gt0.csv 'nonzero(all) > 0' raw_salmon_cts_day3_matrix.csv

abundant <- read_csv('../../11_salmon_quant/data/raw_salmon_cts_day3_matrix_nonzero_all_gt0.csv') %>%
    mutate(Name = str_replace_all(string = Name, pattern = '\\..*$', replacement = '')) %>%
    pull(Name)
  
```

## Get the introns of expressive genes

```{r}
# use the abundant to subset the all_introns 
RET <- all_introns %>%
  filter(ensembl_gene_id %in% abundant)
```

## Write out

How many intron locus all together?

```{r}

RET %>%
  nrow
  
# write out the data
RET %>%
  write_csv('../data/unique_introns_for_expressive_genes.csv')
```

## A look at the table

```{r}
RET %>%
  head() %>%
  kable(caption = 'Table format for the unique introns for the expressive genes') %>%
  kable_styling(full_width = F)
```

