---
title: "The five sets"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
---

# Day 3

## Plan 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
library(tidyverse)
library(kableExtra)
library(ggpubr)
```

```{r out.width=400}
knitr::include_graphics('diagram.jpg')
```

## The five sets

##### The first three: filtering on PSI_KD or PSI_control between 0.1 and 0.9

```{r the first three}
# only looking at the individual PSI values and filter(the first three sets)
# yes finally figured out how to align the IRfinder raw result with the GTF intron result. 
data_path <- '../../05_IRfinder/data/rawIRfinderResult/'
fns <- list.files(data_path, pattern = 'txt', full.names = T)

res <- fns %>%
  map(~ read_tsv(.,comment = '#',col_names = T)  %>%
        mutate(Start = Start + 1) %>%
        mutate(locus = paste(Chr,paste(Start, End, sep = '-'), sep = ':' ) ) %>%
        rename_all(funs(stringr::str_replace_all(., '-', '_') )) %>%
        separate(names(.)[4], into = c('Symbol', 'GeneID', 'Category'), sep = '/') %>%
        mutate(GeneID = str_replace_all(string = GeneID, pattern = '\\..*$', replacement = '')) %>%
        dplyr::select(locus, GeneID, A_IRratio, B_IRratio) %>%
        arrange(desc(B_IRratio))) 


# check if the B_IRratio are the same(should be)
look <- res[[1]] %>%
  inner_join(res[[2]], by = 'locus') %>%
  filter(B_IRratio.x != B_IRratio.y) %>%
  nrow
# ok they are the same
```

```{r}
# get the first three sets locus
set14 <- res[[1]] %>%
  filter(A_IRratio > 0.1 & A_IRratio < 0.9) %>%
  pull(locus)

set89 <- res[[2]] %>%
  filter(A_IRratio > 0.1 & A_IRratio < 0.9) %>%
  pull(locus)

setControl <- res[[2]] %>%
  filter(B_IRratio > 0.1 & B_IRratio < 0.9) %>%
  pull(locus)

list(set14 = set14, set89 = set89, setControl = setControl) %>%
  map_dfr(length) %>%
  kable() %>%
  kable_styling(full_width = F, position = "left")
```

##### The latter two: filtering on padj < 0.05

No PSI filtering since we wanna keep those potentially interesting events. We are not limited to events that we can identify in the lab here only.


```{r}
# the other two sets
# get the sig events of 14 and 89
sig <- fns %>%
  map(~ read_tsv(.,comment = '#',col_names = T)  %>%
        mutate(Start = Start + 1) %>%
        mutate(locus = paste(Chr,paste(Start, End, sep = '-'), sep = ':' ) ) %>%
        rename_all(funs(stringr::str_replace_all(., '-', '_') )) %>%
        separate(names(.)[4], into = c('Symbol', 'GeneID', 'Category'), sep = '/') %>%
        mutate(GeneID = str_replace_all(string = GeneID, pattern = '\\..*$', replacement = ''),
               padj = p.adjust(p_diff, method = 'BH')) %>%
        filter(! Category == 'known-exon') %>%
        filter( A_IRok != 'MinorIsoform' &  B_IRok != 'MinorIsoform' ) %>%
        filter(padj < 0.05) 
        )

# how many of them left
sig %>%
  map_dfc(nrow) %>%
  rename(sig14 = names(.)[1],
         sig89 = names(.)[2]) %>%
  kable() %>%
  kable_styling(full_width = F, position = 'left')
```

```{r}
# extract the sig locus only
sig_locus <- sig %>%
  map(~ pull(.data = ., locus)) 

# get the intersection and union of the sig locus in the above 14 and 89
set_overlap <- intersect(sig_locus[[1]], sig_locus[[2]])
set_union <- union(sig_locus[[1]], sig_locus[[2]])

# make the numbers in to a df
list(set_overlap = set_overlap, set_union = set_union) %>%
  map_dfc(length)  %>%
  kable()  %>%
  kable_styling(full_width = F, position = 'left')
```

## Compared to background introns

##### Keep the introns of the genes that have expression in at least one of the samples and the intron length should be greater than 30.

How many are there?

```{r}
# load the background introns
# filter out ones that have intronLen <= 30, since they are deduced from sequences, could simply be indels 
bg <- read_csv('../data/unique_introns_for_expressive_genes.csv') %>%
  filter(intronLen > 30)

# a summary look at how many introns per gene
p <- bg %>%
  group_by(ensembl_gene_id) %>%
  summarise(cnt = n()) %>%
  gghistogram(x = 'cnt', bins = 100,
              title = 'Histogram of total number of intron locus per gene',
              fill = 'gray')

# how many locus in bg
total_bg <- nrow(bg)
total_bg
```

##### The XXX_Perc column is computed as dividing the following two: 1. the original numbers in each set and 2. the number of locus that also exist in the background intron locus, by the total number specified below respectively.

```{r}
# get the overlap between the five sets and the bg 
ret <- list(set14 = set14,
     set89 = set89,
     setControl = setControl,
     set_overlap = set_overlap,
     set_union = set_union) %>%
  map(~ intersect(., bg$locus))

# the numbers
ret %>%
  map_dfc(length) %>%
  gather(key = Group, value = overlap_num ) %>%
  mutate(overlap_num_Perc = paste(round(overlap_num/total_bg*100, 2), '%', sep = ''))  %>%
  mutate(original_num = list(set14 = set14,
                             set89 = set89,
                             setControl = setControl,
                             set_overlap = set_overlap,
                             set_union = set_union) %>%
                          map_int(length)) %>%
  mutate(original_num_Perc = paste(round(original_num/total_bg*100, 2), '%', sep = ''))  %>%
  kable(caption = 'Percentage of the IR events compared to the background ones') %>%
  kable_styling(full_width = T, position = "left")

```
