---
title: "Smallest intron"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
---

**Are the significantly retained introns the smallest intron in each gene?**

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
library(tidyverse)
library(kableExtra)
library(ggpubr)
```
 
## Create sig RI events

```{r}
# load the sig RI locus in day 3 and check out the length of that RI
# IRfinder : 0 based for start and end
# GTF : 1 based for start and end

# also here I don't have any filtering on PSI or delta PSI
d3RI <- read_csv('../../13_fig1/output/all_tools_sig05_locus_info_day3_with_all_PSI_values.csv') %>%
  dplyr::select(GeneID, locus, type, external_gene_name) %>%
  filter(type == 'RI') %>%
  separate(locus, into = c('chr','coord'), sep = ':', remove = F) %>%
  separate(coord, into = c('start','end'),sep = '-', convert = T) %>%
  mutate(RIlen = end - start)


# collapse the df to see the genes that have more than one RI event
sig <- d3RI %>%
  group_by(GeneID) %>%
  arrange(RIlen, .by_group = TRUE) %>%
  summarise(RI_len_all = paste0(RIlen, collapse = ","),
            locus_all = paste0(locus, collapse = ","),
            cnt = n()) %>%
  arrange(desc(cnt)) %>%
  rename(ensembl_gene_id = GeneID) %>%
  mutate(ensembl_gene_id = str_replace_all(string = ensembl_gene_id, pattern = '\\..*$', replacement = ''))
```

```{r}
# load the data for the background introns
# the genes are the ones that have expression in all 9 samples
# filter out ones that have intronLen <= 30, since they are deduced from sequences, could simply be indels 
bg <- read_csv('../data/unique_introns_for_expressive_genes.csv') %>%
  filter(intronLen > 30)
```

```{r}
# select the shortest intron in each gene
bg_short <- bg %>%
  group_by(ensembl_gene_id) %>%
  summarise( shortest = min(intronLen)) 

# join that with the sig table
res  <- sig %>%
  dplyr::select(ensembl_gene_id, RI_len_all, locus_all) %>%
  left_join(bg_short , by = 'ensembl_gene_id') %>%
  mutate(min_sig_len = str_replace_all(string = RI_len_all, pattern = '\\,.*$', replacement = ''), 
         min_sig_len= as.numeric(min_sig_len) )  %>%
  mutate(abs_len_diff = abs(min_sig_len - shortest))
```


```{r}
res %>%
  head() %>%
  kable(caption = 'A look at the joined table') %>%
  kable_styling()
```


```{r}
res %>%
  split(is.na(.$shortest)) %>%
  map_dfc(nrow) %>%
  rename(non_NA_rows = names(.)[1],
         NA_rows = names(.)[2]) %>%
  kable() %>%
  kable_styling(full_width = F)

# show the distribution of the non-NA-rows
res %>%
  filter(!is.na(shortest)) %>%
  gghistogram(x = 'abs_len_diff', bins = 200,
              color = 'red',
              xlab = 'absolute length difference',
              title = 'Distribution of absolute length difference\nbetween the shortest intron in the gene and the\nshortest retained intron identified by IRfinder') +
  ggsave('../figs/Distribution of absolute length difference.jpg', width = 8, height = 6)
```

```{r out.width=600}
# what is happening at the abs_len_diff > 6000?
# add screenshots form IGV. they look real from the annotation file
file_path <- list.files('../figs/IGV_screenshots _for_long_sig_RI/', full.names = T)
knitr::include_graphics(file_path)
```


## Zoom in 

```{r}
# a zoom in to look at only the abs_len_diff < 500
res %>%
  filter(!is.na(shortest)) %>%
  filter(abs_len_diff < 500) %>%
  gghistogram(x = 'abs_len_diff', bins = 200,
              color = 'red',
              xlab = 'absolute length difference',
              title = 'Same plot but absolute length difference less than 500') +
  ggsave('../figs/Distribution of absolute length difference less than 500.jpg', width = 8, height = 6)
```

```{r}
# a table look at the interval of width 20
interval20 <- res %>%
  filter(!is.na(shortest)) %>%
  filter(abs_len_diff < 500) %>%
  pull(abs_len_diff)  %>%
  cut_interval(., length=20) %>%
  table %>%
  as_tibble %>%
  rename(interval = names(.)[1],
         cnt = names(.)[2])

# group by the abs_len_diff and then do a bar plot
res %>%
  filter(!is.na(shortest)) %>%
  group_by(abs_len_diff) %>%
  summarise(cnt = n()) %>%
  filter(abs_len_diff < 20) %>%
  ggbarplot(x = 'abs_len_diff', y = 'cnt',
            color = 'red',label = T,
            title = 'Tally of absolute len difference when it is below 20',
            xlab = 'absolute length difference')  +
  ggsave('../figs/Tally of absolute len difference when it is below 20.jpg', width = 6, height = 5)

res %>%
  filter(!is.na(shortest)) %>%
  group_by(abs_len_diff) %>%
  summarise(cnt = n()) %>% 
  mutate(thre20 = abs_len_diff < 20) %>% 
  split(.$thre20) %>%
  map_dfc( ~ sum(.$cnt)) %>%
  rename(gt20 = names(.)[1],
         lt20 = names(.)[2]) %>%
  kable(caption = 'How many abs_len_diff < 20 and how many are > 20') %>%
  kable_styling( position = 'left')

```

Of all the genes that have sig RI events, what percentage of them has a RI event that is the smallest intron in that gene? (One gene may have multiple RI events. If there exist one event that is the smallest intron in the gene, it will be counted. )

```{r}
res %>%
  filter(abs_len_diff == 0) %>%
  summarise(Num = nrow(.)) %>%
  mutate(Total = nrow(res)) %>%
  mutate(Percentage = paste(round(Num/Total*100, 2), '%', sep = '')) %>%
  kable(caption = 'what percentage of geneshas a RI event that is the smallest intron in that gene?') %>%
  kable_styling(full_width = F, position = 'left')
  
  
```



## Sig but not abundant

```{r}
# rows where shortes is NA(doesn't make into the bg list) will check later
NA_rows <- res %>%
  filter(is.na(shortest))


NA_rows %>%
  kable(caption = "Gene that doesn't exist in the background") %>%
  kable_styling


  
```


