---
title: "PSI14 and 89"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
---

Day 3.

Jess: a list of non-sig genes with delta PSI and individual PSIs Similar to the document "Day3 RNAseq IR list from all tools" -- going to pick a non-significantly retained intron to show there is no difference in the non-sig gene list between control and knockdown. 


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
library(tidyverse)
library(ggpubr)
library(kableExtra) 
```

Get the non-significant results in each tool.

How to get the intersection in each tool:

+ Get the non-significant results in 14 and 89 respectively
+ Take the intersection 

# Whippet intersection

```{r}
# data path for whippet
data_path <- '../../03_ASprofile/data/whippetResult'
fns <- list.files(data_path, pattern = 'output.diff', full.names = T)

# read the data and select the columns
whippet_sig <- fns %>%
  map(~ read_tsv(.,col_names = T)  %>%
        mutate(absDeltaPsi = abs(DeltaPsi)) %>%
        filter(absDeltaPsi <= 0.1 & Probability > 0.9 & Psi_A > 0.1) %>%
        mutate(tool = 'Whippet') %>%
        dplyr::select(ensembl_gene_id = Gene, 
                      locus = Coord, 
                      type = Type,
                      Psi_A, 
                      Psi_B,
                      tool) %>%
        mutate(ensembl_gene_id = str_replace_all( ensembl_gene_id, '\\..*$', ''))
   )

# find the intersection by inner join
whippet_ol  <- whippet_sig[[1]] %>%
  rename(Psi_14 = Psi_A) %>%
  inner_join(whippet_sig[[2]] %>%
               dplyr::select(Psi_89 = Psi_A, locus), by = 'locus')
        
```

# IRfinder intersection

```{r}
data_path <- '../../05_IRfinder/data/rawIRfinderResult/'
fns <- list.files(data_path, pattern = 'txt', full.names = T)

IRfinder_sig <- fns %>%
  map(~ read_tsv(.,comment = '#',col_names = T)  %>%
        mutate(Start = Start + 1) %>%
        mutate(locus = paste(Chr,paste(Start, End, sep = '-'), sep = ':' ) ) %>%
        rename_all(funs(stringr::str_replace_all(., '-', '_') )) %>%
        separate(names(.)[4], into = c('Symbol', 'GeneID', 'Category'), sep = '/') %>%
        mutate(ensembl_gene_id = str_replace_all(string = GeneID, pattern = '\\..*$', replacement = ''),
               padj = p.adjust(p_diff, method = 'BH')) %>%
        filter(! Category == 'known-exon') %>%
        filter(A_IRok != 'MinorIsoform' &  B_IRok != 'MinorIsoform' ) %>%
        filter(padj >= 0.05)  %>%
        mutate(type = "RI",
               tool = 'IRfinder') %>%
        dplyr::select(ensembl_gene_id, 
                      locus,
                      type,
                      Psi_A = A_IRratio, 
                      Psi_B = B_IRratio,
                      tool))

IRfinder_ol <- IRfinder_sig[[1]] %>%
  rename(Psi_14 = Psi_A) %>%
  inner_join(IRfinder_sig[[2]] %>%
               dplyr::select(Psi_89 = Psi_A, locus), by = 'locus')
```

# rMATS not included

Because they are providing too many results like 70,000+ 


```{r}
# how many of them each
tibble(
  IRfinder_ol = nrow(IRfinder_ol),
  whippet_ol = nrow(whippet_ol)
) %>%
  kable(caption = 'Number of overlapped non-sig events between 14 and 89 in each tool') %>%
  kable_styling(full_width = F, position = 'left')
```


# Combine three

+ Combine all the significant results from three tools together, and remove duplicated locus.
+ PSI_KD is the mean of Psi_14 and Psi_89.
+ PSI_KD - PSI_WT = deltaPSI
+ The table is sorted in the descending order of abs(deltaPSI)

```{r}
all_sig <- bind_rows(
  IRfinder_ol,
  whippet_ol
) %>%
  distinct(locus, .keep_all = TRUE) 
```

```{r}
# calculate the mean of 14 and 89 as the PSI_KD and
# PSI_KD - PSI_B = deltaPSI
ret <- all_sig %>%
  rowwise() %>%
  mutate(PSI_KD = mean(c(Psi_14, Psi_89), na.rm = T)) %>%
  rename(PSI_WT = Psi_B) %>%
  mutate(deltaPSI = PSI_KD - PSI_WT) %>%
  arrange(desc(abs(deltaPSI)))
```

```{r}
# add the gene name column to the table
library(biomaRt)

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# the ensembl_gene_id of the genes that you are interested in 
genes <- ret %>%
  distinct(ensembl_gene_id) %>%
  pull(ensembl_gene_id)

# retrieve the feature information of the listed attributes
symbol <- getBM(filters = "ensembl_gene_id",
                attributes = c('ensembl_gene_id','external_gene_name'),
                values = genes, 
                mart = mart)
```

```{r}
# join the gene name to the ret table
ret <- ret %>%
  left_join(symbol, by = 'ensembl_gene_id')

nrow(ret)
```


```{r}
# write out the above table . this is the result of sig events not filtered by PSI or deltaPSI

ret %>%
  dplyr::select(ensembl_gene_id,
                gene_name = external_gene_name,
                locus,
                type,
                Psi_14, Psi_89, PSI_KD, PSI_WT,deltaPSI,
                tool) %>% 
  write_csv('../output/day3_non_sig_events_2_tools_not_filtered_by_PSI.csv')
```

