---
title: "overlap_with_geneset"
author: "Angel"
date: "1/9/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(ggpubr)
library(knitr)
```

```{r}
# the function to do the fisher
calcHyper <- function(geneSet,sig,allGenes){

   a <- length(intersect(sig,geneSet))
   b <- length(geneSet)-a
   c <- length(sig)-a
   d <- allGenes-a-b-c
   cont <- matrix(c(a,b,c,d),ncol=2)
   pVal <- fisher.test(cont)
   return(c(overlap=a,geneset=a+b,sig=c+a,background=allGenes,overlap_per = round(a/(a+b),digits = 3)*100))
}
```


```{r}
# background genes
eset <- read_rds('../../08_findingGenes/data/Norm_salmon_cts_replicate1_eset_unfiltered.RDS')
allGenes <- nrow(fData(eset))
background <- fData(eset) %>%
  rownames_to_column 
  

# the sig genes 
sig <- read.table('../../08_findingGenes/data/union_sig_genes_from_3_analysis.txt', stringsAsFactors = F) %>%
  rename(rowname = V1) %>%
  left_join(background, by = 'rowname') %>%
  pull(external_gene_name)
  

# the non sig but have expression genes and have at least three introns
blue3 <- read_csv('../../09_mitochondria/output/genes_with_at_least_3_introns_ids.txt', col_names = F) %>%
  rename(rowname = X1) %>%
  left_join(background, by = 'rowname') %>%
  pull(external_gene_name)
```

```{r}
# the msigDB geneset 
path <- '../../../../../../undertaking/bioinformatics_container/MsigDB_geneset/symbol_6.2'
gmt_short <- list.files(path, pattern = 'gmt')
gmts <- list.files(path, pattern = 'gmt', full.names = T)

Gset_list <- lapply(gmts, function(gmt){
  gsets = readLines(gmt)
  gsets = strsplit(gsets,'\t')
  names(gsets) = sapply(gsets,function(x)x[1])
  gsets = sapply(gsets,function(x)x[-(1:2)])
  return(gsets)
})

names(Gset_list) <- gmt_short
length(Gset_list)
```

```{r overlap_sig}
RES <- list()


for (i in 1:length(Gset_list)) {
  RES[[i]] = t(sapply(Gset_list[[i]],calcHyper,sig,allGenes)) %>%
    as.data.frame %>%
    rownames_to_column %>%
    rename(gset_name = rowname) %>%
    mutate(collection = names(Gset_list)[i]) %>%
    arrange(desc(overlap_per))
}

sig_overlap_msigdb <- bind_rows(RES)
  
```


```{r overlap_blue3}
RES <- list()


for (i in 1:length(Gset_list)) {
  RES[[i]] = t(sapply(Gset_list[[i]],calcHyper,blue3,allGenes)) %>%
    as.data.frame %>%
    rownames_to_column %>%
    rename(gset_name = rowname) %>%
    mutate(collection = names(Gset_list)[i]) %>%
    arrange(desc(overlap_per))
}

blue3_overlap_msigdb <- bind_rows(RES)
  
```

