---
title: "Multiple sequence alignment in introns of comparable lengths in sig and non-sig"
author: "Anqi Dai"
date: "6/18/2019"
output: html_document
---
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(BSgenome)
library(BSgenome.Hsapiens.NCBI.GRCh38)
library(seqinr)
library(tidyverse)
```


```{r}
# reformat the function so that the function's job is to take a df as input and return a df that has an additinal column as the sequences that got retrieved
get_seq_from_df <- function(df){
  genome <- BSgenome.Hsapiens.NCBI.GRCh38
  seqlevelsStyle(genome) = "UCSC"

  grange = makeGRangesFromDataFrame(df, 
                                  seqnames.field=c("chr"), 
                                  start.field=c("start"),
                                  end.field=c("end"), 
                                  strand.field=c("direction"))
  seq = BSgenome::getSeq(genome, grange)
   
  ret = df %>%
    mutate(sequence = as.character(seq)) %>% 
    mutate(GC = stringr::str_count(sequence, "G") + stringr::str_count(sequence, "C")) %>%
    mutate(GC_per = GC / stringr::str_length(sequence) * 100)
  return(ret)
}
```

```{r}
fns <- list.files('../output', pattern = 'under', full.names = T)
out <- fns %>% 
  set_names(fns) %>% 
  map(~ read_csv(.)) %>% 
  map(~ get_seq_from_df(.))
  
out_seq <- out %>% 
  imap(~ mutate(.data = ., fn = .y) %>% 
         mutate(id = seq(1, nrow(.))) %>% 
         dplyr::select(sequence, id, fn))
       
# so the write.fasta function that writes out the fasta do not fit well into a pipe. it has to be done separately
for(i in 1:length(out_seq)) {
  df = out_seq[[i]]
  fn = unique(df$fn)
  write.fasta(sequences = as.list(df$sequence), 
              names = df$id,
              file.out = str_glue('{fn}.fa'))
}

```

