---
title: "retrieve sequence info for AS and non AS 500 seqs"
author: "Anqi Dai"
date: "6/26/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(BSgenome)
library(BSgenome.Hsapiens.NCBI.GRCh38)
library(seqinr)
library(tidyverse)
```

* Get all of the sig and non-sig sequences under len 300.
* Sort by len ascending order and get the top 500 (since Muscle can only take 500 sequences maximum)
* Get all of the non AS short sequences below the len of the previous step
* Get the two groups sequences and its GC content.

```{r}
# get the sig and non sig separatly for two comparisons.

data_path <- '../../05_IRfinder/data/rawIRfinderResult/'
fns <- list.files(data_path, pattern = 'txt', full.names = T)

sig <- fns %>%
  map(~ read_tsv(.,comment = '#',col_names = T)  %>%
        mutate(Start = Start + 1) %>%
        mutate(locus = paste(Chr,paste(Start, End, sep = '-'), sep = ':' ) ) %>%
        rename_all(funs(stringr::str_replace_all(., '-', '_') )) %>%
        separate(names(.)[4], into = c('Symbol', 'GeneID', 'Category'), sep = '/') %>%
        mutate(GeneID = str_replace_all(string = GeneID, pattern = '\\..*$', replacement = ''),
               padj = p.adjust(p_diff, method = 'BH')) %>%
        filter(! Category == 'known-exon') %>%
        filter(A_IRok != 'MinorIsoform' &  B_IRok != 'MinorIsoform' ) %>%
        filter(padj < 0.05) 
      )

non_Sig <- fns %>%
  map(~ read_tsv(.,comment = '#',col_names = T)  %>%
        mutate(Start = Start + 1) %>%
        mutate(locus = paste(Chr,paste(Start, End, sep = '-'), sep = ':' ) ) %>%
        rename_all(funs(stringr::str_replace_all(., '-', '_') )) %>%
        separate(names(.)[4], into = c('Symbol', 'GeneID', 'Category'), sep = '/') %>%
        mutate(GeneID = str_replace_all(string = GeneID, pattern = '\\..*$', replacement = ''),
               padj = p.adjust(p_diff, method = 'BH')) %>%
        filter(! Category == 'known-exon') %>%
        filter(A_IRok != 'MinorIsoform' &  B_IRok != 'MinorIsoform' ) %>%
        filter(padj >= 0.05) 
      )
```
 
```{r}
# now the sig locus is the intersection
# and the nonSig is still the intersection
sig_locus <- intersect(sig[[1]]$locus, sig[[2]]$locus)

non_Sig_locus <- intersect(non_Sig[[1]]$locus, non_Sig[[2]]$locus)
 
tibble(sig_locus = length(sig_locus),
       non_Sig_locus = length(non_Sig_locus)) 
```
  
```{r}
# getting the length of the locus
sig_df <-  sig[[1]] %>%
  filter(locus %in% sig_locus) %>%
  dplyr::select(chr = Chr, start = Start, end = End, direction = Direction)  %>%
  mutate(RIlen = end - start,
         category = 'Sig') %>%
  arrange(RIlen)  

non_Sig_df <-  non_Sig[[1]] %>%
  filter(locus %in% non_Sig_locus) %>%
  dplyr::select(chr = Chr, start = Start, end = End, direction = Direction)  %>%
  mutate(RIlen = end - start,
         category = 'Non-sig') %>%
  arrange(RIlen)  
 
```

```{r}
# get all of the sig and non-sig sequences under len 300 and sort by len ascending order and get the top 500
AS_thre <- bind_rows(sig_df, non_Sig_df) %>% 
  filter(RIlen < 300) %>% 
  arrange(RIlen) 

N <- nrow(AS_thre)
 
AS_thre_N <- AS_thre[1:N,] 

```

```{r}
# Get all of the non AS short sequences below the len of the previous step
len_thre_AS <- max(AS_thre$RIlen)


nashort_thre <- read_csv('../../05_IRfinder/data/03_non_AS_seq_info.csv')  %>% 
  filter(intronLen < 358) %>% 
  dplyr::select(direction = strand, locus, RIlen = intronLen, category) %>% 
  separate(col = locus, into = c('chr','coord'), sep = ':', remove = T) %>%
  separate(coord, into = c('start','end'), sep = '-', convert = T) %>%
  dplyr::select(chr:end, direction, RIlen:category) %>% 
  filter(RIlen < len_thre_AS) %>% 
  arrange(RIlen)
```
```{r}
# get the sequences and the GC content for the above two tables
# reformat the function so that the function's job is to take a df as input and return a df that has an additinal column as the sequences that got retrieved
get_seq_from_df <- function(df){
  genome <- BSgenome.Hsapiens.NCBI.GRCh38
  seqlevelsStyle(genome) = "UCSC"

  grange = makeGRangesFromDataFrame(df, 
                                  seqnames.field=c("chr"), 
                                  start.field=c("start"),
                                  end.field=c("end"), 
                                  strand.field=c("direction"))
  seq = BSgenome::getSeq(genome, grange)
   
  ret = df %>%
    mutate(sequence = as.character(seq)) %>% 
    mutate(GC = stringr::str_count(sequence, "G") + stringr::str_count(sequence, "C")) %>%
    mutate(GC_per = GC / stringr::str_length(sequence) * 100)
  return(ret)
}
```

```{r}
# run the above function on nashort_thre and AS_thre_N
# combine the gc and RIlen to make a new column
out <- list(nonas = nashort_thre,
            as = AS_thre) %>% 
  map(~ get_seq_from_df(.) %>% 
        dplyr::select(RIlen, sequence, GC_per) %>% 
        mutate(id = seq(1, nrow(.)),
               combi = paste(RIlen, floor(GC_per), sep = '-'))) 


# filter the nonas by that combi first
unique_combi <- unique(out$as$combi)
  
nonas_fil <- out$nonas %>% 
  filter(combi %in% unique_combi) 


```

```{r}
set.seed(456)
# get 1000 samples of the non-AS since 100 samples won't resolve.
sample_times <- 1000
n_samples <- rerun(sample_times, nonas_fil %>% 
                     sample_n(N))
names(n_samples) <- seq(1,sample_times)

# write out to disk which will do the msa later
for(i in 1:length(n_samples)) {
  df = n_samples[[i]]
  name = names(n_samples)[i]
  write.fasta(sequences = as.list(df$sequence), 
              names = df$id,
              file.out = str_glue('../output/s{name}_gc_len_comparable_sampling_len_lt_300_all.fa'))
}
```


