---
title: "Reading the alignments in fasta format"
author: "Anqi Dai"
date: "7/2/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(seqinr)
library(tidyverse)
library(ggpubr)
library(kableExtra)
```

The shadedness is the percentage of the dominant nulcetide among all the base in that locus.

```{r}
fasta.res  <- read.alignment(file = '../output/entropy_as_all.fa',
 format = "fasta")

# how long is each string
len <- nchar(fasta.res$seq[[1]])
  

res <- fasta.res$seq %>% 
  map(function(str){
  str_split_fixed(str, pattern = '',
                  n = len) %>% 
      as.data.frame(stringsAsFactors = FALSE )
})

ret <- res %>% 
  bind_rows


# converting it to numeric values : -:0, A:1, G:2, C:3, T:4

converting_to_numeric <- function(column){
  case_when(
    column == '-' ~ 0,
    column == 'a' ~ 1,
    column == 'g' ~ 2,
    column == 'c' ~ 3,
    column == 't' ~ 4,
  )
}

num <- apply(ret, 2, converting_to_numeric)

# 0 doesn't engage in any calculations!
# out of the all of the nonzero digits, what is the percentage of the dominant digit? in each column?

# find out the nonzero digit in each column
total_nonzeros <- colSums(num != 0) %>% 
  as.data.frame(stringsAsFactors = FALSE) %>% 
  rownames_to_column('colu_name') %>% 
  rename(total_non0 = names(.)[2])


domninant <- apply(num, 2, function(colu){
  ret = table(colu) %>% 
    as.data.frame(stringsAsFactors = FALSE ) %>% 
    filter(colu != 0) %>% 
    filter(Freq == max(Freq)) %>% 
    head(1) 
  return(ret)
}) 


RES <- domninant %>% 
  bind_rows %>% 
  mutate(colu_name = names(domninant))

# use a series of threshold to decide which cutoff to choose as there are enough nt aligning at that locus
all <- RES %>% 
  left_join(total_nonzeros, by= 'colu_name') %>% 
  arrange(desc(Freq)) %>% 
  mutate(total_seq = fasta.res$nb) %>% 
  mutate(perc_nonzero = total_non0/total_seq)

# bind that together to make a df
cutoffs <- seq(0.6, 0.9, 0.05) 

cutoff <- cutoffs %>% 
  set_names(cutoffs) %>% 
  map_dfc(~ filter(.data = all, perc_nonzero > .) %>% 
          nrow(.)) %>% 
  gather(key = threshold, value = number)


cutoff %>% 
  kable(caption = 'the number of surviving locus after filtering the percentage of non-gap sequences') %>% 
  kable_styling(full_width = F, position = 'left')
```

```{r}

# now join them together to have a table with both total number of nonzeros and the number of dominant digits
# totalnonzero should be at least 80% of the total sequences
cutoffs %>% 
  map(~ filter(.data = all, perc_nonzero > .) %>% 
  mutate(perc_domi = round(Freq/total_non0, 2),
         nt = factor(ifelse(colu == 2, 'G', ifelse(colu == 3, 'C', ifelse(colu == 4, 'T', 'A'))))) %>% 
  ggplot(aes(x = nt, y = perc_domi)) +
  geom_jitter(width = 0.1, color = '#EC0000', size = 1) +
  theme_light() +
  labs(x = 'Nucleotide',
       y = 'Fraction of the dominant nucleotide',
       title = str_glue('Fraction of non-gap sequences: threshold {.}')) +
    ggsave(str_glue('../figs/AS/boxplot perc of the dom nucleotide- cutoff {.}.jpg'), width = 5, height =4, dpi = 300)) 

```

