---
title: "Compare novel and normal quantification in 14 VS control"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
---

Compare the counts for genes when considering annotated transcriptom and also the one includes the intron. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning = FALSE)
library(Biobase)
library(tidyverse)
library(ggpubr)
library(kableExtra)
library(biomaRt)
```


```{r}
novel14 <- read_tsv('../output/novel_14_control_DE.csv')
normal14 <- read_tsv('../output/normal_14_control_DE.csv')

res14 <- novel14 %>%
  dplyr::select(X, novel14log2FC = Status__KD1__log2FoldChange) %>%
  inner_join(normal14 %>%
              dplyr::select(X, normal14log2FC = Status__KD1__log2FoldChange), by = 'X')
```


```{r}
res14 <- novel14 %>%
  dplyr::select(X, novel14log2FC = Status__KD1__log2FoldChange, novel_baseMean = baseMean) %>%
  inner_join(normal14 %>%
              dplyr::select(X, normal14log2FC = Status__KD1__log2FoldChange, normal_baseMean= baseMean), by = 'X') 
```


```{r eval=F}
# the correlation between the basemean of the two quant
cor.test(res14$novel_baseMean,res14$normal_baseMean, method = 'spearman')
```

## log2FC comparison

```{r}
# color the dots with red: AS blue: not AS but have expression
sig_gene <- read_table('../../08_findingGenes/data/union_sig_genes_from_3_analysis.txt', col_names = F)

detkF <- read_csv('../../08_findingGenes/data/detk_fil_Norm_salmon_cts_replicate1_matrix.csv') %>%
  mutate(Name = str_replace_all(string = Name, pattern = '\\..*$', replacement = ''))

blue_genes <- setdiff(detkF$Name, sig_gene$X1)

red_genes <- intersect(detkF$Name, sig_gene$X1)

```

```{r}
# mark the res14 table with the different categories of genes


res <- res14 %>% 
  mutate(Color = factor(if_else(X %in% blue_genes, 'no AS', if_else(X %in% red_genes, 'AS', 'other'))) ) %>%
  mutate(novelNormalDiffFC = novel14log2FC - normal14log2FC ) %>%
  filter(abs(normal14log2FC) < 5) %>%
  filter(abs(novel14log2FC) < 5) 
```

```{r}
# plot the multiple colors
res %>%
  ggplot(aes(x = novelNormalDiffFC, fill = Color, color  = Color )) +
  geom_histogram(bins = 80) +
  geom_vline(xintercept = 0, color = 'white', alpha = 1) + 
  labs(x = 'novel14log2FC - normal14log2FC',
       title = 'Diff in novel & normal quant log2FC in histogram') +
  facet_wrap(.~ Color) + 
  scale_fill_manual(values =  c('#EC0000', '#00468B')) +
  scale_color_manual(values =  c('#EC0000', '#00468B'))  +
  ggsave('../figs/Diff in novel & normal quant log2FC in histogram 14.jpg', width = 8, height =6, dpi = 300)
```

```{r}
# the log2FC in scatter
res %>%
  ggplot(aes(x = normal14log2FC, y = novel14log2FC, color = Color))+
  geom_point(alpha = 0.2) +
  geom_abline(slope=1, intercept=0, color = 'black', alpha = 0.5) +
  labs(x = 'log2FC in normal quantification', 
       y ='log2FC in novel quantification',
       title = 'Novel VS normal log2FC(removed extreme outlier)') + 
  scale_fill_manual(values =  c('#EC0000', '#00468B')) +
  scale_color_manual(values =  c('#EC0000', '#00468B'))  +
  ggsave('../figs/novel_normal_log2FC_color 14.jpg', width = 8, height =6, dpi = 300)
```

```{r eval=F}
# see where those 800 genes are in the plot only 
high <- read_table('../../09_mitochondria/output/genes_with_at_least_3_introns_ids_highTPM.txt', col_names = F)

res %>%
  mutate(high = factor(ifelse(X %in% high$X1, 'Blue highTPM', 'other'))) %>%
  filter(high == 'Blue highTPM') %>%
  ggplot(aes(x = normal14log2FC, y = novel14log2FC, shape = high))+
  geom_point(alpha = 0.3) +
  geom_abline(slope=1, intercept=0, color = 'black', alpha = 0.5) +
  labs(x = 'log2FC in normal quantification', 
       y ='log2FC in novel quantification',
       title = 'The 800 genes that have high TPM but are not AS') + 
  scale_fill_manual(values =  c('#00468B', '#EC0000')) +
  scale_color_manual(values =  c('#00468B', '#EC0000'))  +
  ggsave('../figs/The 800 genes that have high TPM but are not AS 14.jpg', width = 8, height =6, dpi = 300)
```

## Basemean comparison

```{r}
# the basemean plot with colors
res %>%
  ggplot(aes(x = normal_baseMean, y = novel_baseMean, color = Color))+
  geom_point(alpha = 0.2) +
  yscale("log10", .format = TRUE)+
  xscale("log10", .format = TRUE) +
  geom_abline(slope=1, intercept=0, color = 'black', alpha = 0.5) +
   labs(title = 'Novel VS normal basemean') +
  scale_fill_manual(values =  c('#EC0000', '#00468B')) +
  scale_color_manual(values =  c('#EC0000', '#00468B')) +
  ggsave('../figs/novel_normal_basemean_color 14.jpg', width = 8, height =6, dpi = 300)
```


Those blue points above the line represent they have much higher base mean when you map the reads to the whole gene (including the intronic region), so the question is why are they identified AS genes by IRfinder?

*Select some of those genes*

novel_baseMean > 10000 & normal_baseMean < 50

```{r}
# select some genes in IRfinder
test <- res %>%
  filter(novel_baseMean > 10000 & normal_baseMean < 50) %>%
  pull(X)

# get the gene info for those genes
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# retrieve the feature information of the listed attributes
symbol <- getBM(filters = "ensembl_gene_id",
                attributes = c('ensembl_gene_id',
                               'external_gene_name',
                               'chromosome_name',
                               'start_position',
                               'end_position', 
                               'description'),
                values = test, 
                mart = mart)
symbol <- symbol %>%
  rename(chr = chromosome_name)

symbol %>%
  kable() %>%
  kable_styling()
```

## log(basemen) vs log(FC)


```{r}
# plot log basemen vs log fold change for AS and non-as
res <- res %>%
  mutate(logBM_novel = log2(novel_baseMean),
         logBM_normal = log2(normal_baseMean))

res %>%
  ggplot(aes(x = novel14log2FC, y = logBM_novel )) +
  geom_point(alpha = 0.1, color = '#00468B') +
  facet_wrap(.~Color)+
  labs(x = 'log2 Fold Change',
       y = 'log2 base mean',
       title = 'novel quantification')


res %>%
  ggplot(aes(x = normal14log2FC, y = logBM_normal )) +
  geom_point(alpha = 0.1, color = '#00468B') +
  facet_wrap(.~Color)+
  labs(x = 'log2 Fold Change',
       y = 'log2 base mean',
       title = 'normal quantification')
```

## Filter out the pseudogenes

```{r}

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# retrieve the feature information of the listed attributes
symbol <- getBM(filters = "ensembl_gene_id",
                attributes = c('ensembl_gene_id',
                               'external_gene_name',
                               'chromosome_name',
                               'start_position',
                               'end_position', 
                               'description'),
                values = res$X, 
                mart = mart)
symbol <- symbol %>%
  rename(chr = chromosome_name)


# join to the res table
ret <- symbol %>%
  dplyr::select(ensembl_gene_id, description) %>%
  left_join(res %>%
              rename(ensembl_gene_id = X), by  ='ensembl_gene_id') %>%
  filter(! grepl('pseudogene|novel transcript', description))



```

```{r}
# re do the basemean plot

ret %>%
  ggplot(aes(x = normal_baseMean, y = novel_baseMean, color = Color))+
  geom_point(alpha = 0.2) +
  yscale("log10", .format = TRUE)+
  xscale("log10", .format = TRUE) +
  geom_abline(slope=1, intercept=0, color = 'black', alpha = 0.5) +
   labs(title = 'Novel VS normal basemean(remove pseudogenes)') +
  scale_fill_manual(values =  c('#EC0000', '#00468B')) +
  scale_color_manual(values =  c('#EC0000', '#00468B'))
```

```{r}
ret %>%
  ggplot(aes(x = novelNormalDiffFC, fill = Color, color  = Color )) +
  geom_histogram(bins = 80) +
  geom_vline(xintercept = 0, color = 'white', alpha = 1) + 
  labs(x = 'novel14log2FC - normal14log2FC',
       title = 'Diff in novel & normal quant log2FC in histogram(remove pseudogenes)') +
  facet_wrap(.~ Color) + 
  scale_fill_manual(values =  c('#EC0000', '#00468B')) +
  scale_color_manual(values =  c('#EC0000', '#00468B'))
```

```{r}
ret %>%
  ggplot(aes(x = normal14log2FC, y = novel14log2FC, color = Color))+
  geom_point(alpha = 0.2) +
  geom_abline(slope=1, intercept=0, color = 'black', alpha = 0.5) +
  labs(x = 'log2FC in normal quantification', 
       y ='log2FC in novel quantification',
       title = 'Novel VS normal log2FC(remove pseudogenes)') + 
  scale_fill_manual(values =  c('#EC0000', '#00468B')) +
  scale_color_manual(values =  c('#EC0000', '#00468B'))
```



