---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---
 
```{r}
library(Biobase)
library(tidyverse)
library(ggpubr)
library(kableExtra)
```

## Get the DE result for the novel quantification 

```{r}
# load the raw result from the salmon quant for novel way
novel_raw <- read_tsv('../data/raw_salmon_cts_replicate1_matrix_GENE_feature.csv')
```

```{r}
# replace the locus row ID with ensembl_gene_id
# reformat the sample name


# get the locus and gene ID match information

# thru awk:
#  awk -F "\t" '$3 == "gene" { print $1,$4,$5,$9 }' gencode.v27.annotation.gtf | awk '{ for(i=1;i<=NF;i++){if(i==NF){printf("%s\n",$NF);}else {printf("%s\t",$i)}}}' > gencode.v27.annotation.gene.bed
#  awk -F " "  '{print $1":"$2"-"$3,$5}' gencode.v27.annotation.gene.bed  > gencode.v27.annotation.gene.match.bed  

match <- read.table('../data/gencode.v27.annotation.gene.match.bed', stringsAsFactors = F) %>%
  dplyr::select(Name = V1, ensembl_gene_id = V2) %>%
  mutate(ensembl_gene_id = stringr::str_replace_all(ensembl_gene_id, '\\..+$',''))

novel <- novel_raw %>%
  left_join(match, by = 'Name') %>%
  filter(!duplicated(Name)) %>%
  dplyr::select(-Name) %>%
  column_to_rownames('ensembl_gene_id') %>%
  rename_all(
    funs(stringr::str_replace_all(., '-', '_')))

write.csv(novel,  '../data/novel_quant_salmon_raw_cts.csv', row.names = T, quote = F)
```

```{bash}
# filter out rows that have any zero
# Do this prior to DE
# after this filtering you can directly supply the matrix to the de module that will do the norm work
# detk-filter -o ../data/novel_quant_salmon_raw_cts_filtered.csv 'zero(all) == 0' ../data/novel_quant_salmon_raw_cts.csv
# 20887 rows in the filtered data
# worked! on my mac
```

```{bash}
# de 14 VS control

#detk-de deseq2 "counts ~ Status" ../data/novel_quant_salmon_raw_cts_filtered.csv ../../09_mitochondria/data/pheno14_ctrl.csv > ../output/novel_14_control_DE.csv


# de 89 VS control

#detk-de deseq2 "counts ~ Status" ../data/novel_quant_salmon_raw_cts_filtered.csv ../../09_mitochondria/data/pheno89_ctrl.csv > ../output/novel_89_control_DE.csv
```


##  Get the DE result for the normal quantification 

Have to re-do this because I have to add the filtering step prior to DE

```{r}
# clean the table a little bit
normal <- read_tsv('../data/raw_salmon_cts_replicate1_matrix_Normal.csv') %>%
  mutate(Name = stringr::str_replace_all(Name, '\\..+$','')) %>%
  rename_all(
    funs(stringr::str_replace_all(., '-', '_'))) %>%
  column_to_rownames('Name') 

write.csv(normal,  '../data/normal_quant_salmon_raw_cts.csv', row.names = T, quote = F)
```


```{bash}
# filter on the raw counts matrix
#detk-filter -o ../data/normal_quant_salmon_raw_cts_filtered.csv 'zero(all) == 0' ../data/normal_quant_salmon_raw_cts.csv
# 21361 rows in the filtered data
```

```{r}
# de 14 VS control

#detk-de deseq2 "counts ~ Status" ../data/normal_quant_salmon_raw_cts_filtered.csv ../../09_mitochondria/data/pheno14_ctrl.csv > ../output/normal_14_control_DE.csv

# de 89 VS control

#detk-de deseq2 "counts ~ Status" ../data/normal_quant_salmon_raw_cts_filtered.csv ../../09_mitochondria/data/pheno89_ctrl.csv > ../output/normal_89_control_DE.csv
```

## Compare the log2 Fold Change of those genes in the intersection of novel and normal in 14control and 89control respectively

```{r}
novel14 <- read_tsv('../output/novel_14_control_DE.csv')
normal14 <- read_tsv('../output/normal_14_control_DE.csv')

res14 <- novel14 %>%
  dplyr::select(X, novel14log2FC = Status__KD1__log2FoldChange) %>%
  inner_join(normal14 %>%
              dplyr::select(X, normal14log2FC = Status__KD1__log2FoldChange), by = 'X')
```

```{r plot}
# plot
res14 %>%
  ggplot(aes(x = normal14log2FC, y = novel14log2FC))+
  geom_point(alpha = 0.2) +
  geom_abline(slope=1, intercept=0, color = 'red') +
  labs(x = 'log2FC in normal quantification', 
       y ='log2FC in novel quantification',
       title = 'Novel VS normal log2FC')
```

```{r}
# what is that gene so far away from the others
res14 %>%
  filter(normal14log2FC < -20) %>%
  kable() %>%
  kable_styling(full_width = F)

# ENSG00000283029 is an RNA Gene, and is affiliated with the SRP RNA class.
```

```{r}
# plot again without that gene
res14 %>%
  filter(normal14log2FC > -20) %>%
  ggplot(aes(x = normal14log2FC, y = novel14log2FC))+
  geom_point(alpha = 0.2) +
  geom_abline(slope=1, intercept=0, color = 'red') +
  labs(x = 'log2FC in normal quantification', 
       y ='log2FC in novel quantification',
       title = 'Novel VS normal log2FC(remove extreme outlier)')
```
```{r}
# what are those genes have much higher log2FC in novel than normal
res14 %>%
  filter(abs(novel14log2FC) > 10) %>%
  kable() %>%
  kable_styling(full_width = F)

# Histone Cluster 2 H4 Family Member A . This gene is intronless and encodes a replication-dependent histone that is a member of the histone H4 family. Transcripts from this gene lack polyA tails;

# C17orf49 (Chromosome 17 Open Reading Frame 49) is a Protein Coding gene. Gene Ontology (GO) annotations related to this gene include chromatin binding.

# ENSG00000283515 is a Protein Coding gene.
```

```{r}
res14 %>%
  filter(normal14log2FC > -20) %>%
  filter(abs(novel14log2FC) < 5) %>%
  ggplot(aes(x = normal14log2FC, y = novel14log2FC))+
  geom_point(alpha = 0.1) +
  geom_abline(slope=1, intercept=0, color = 'red') +
  labs(x = 'log2FC in normal quantification', 
       y ='log2FC in novel quantification',
       title = 'Novel VS normal log2FC(remove extreme outlier)')
```

```{r}

res14 %>%
  filter(normal14log2FC > -20) %>%
  filter(abs(novel14log2FC) < 5) %>%
  mutate(Diff = novel14log2FC- normal14log2FC) %>%
  gghistogram(x = 'Diff', xlab = 'novel14log2FC- normal14log2FC',
              ylab= 'Count')
 


```

```{r}
res14 <- novel14 %>%
  dplyr::select(X, novel14log2FC = Status__KD1__log2FoldChange, novel_baseMean = baseMean) %>%
  inner_join(normal14 %>%
              dplyr::select(X, normal14log2FC = Status__KD1__log2FoldChange, normal_baseMean= baseMean), by = 'X') 

cor.test(res14$novel_baseMean,res14$normal_baseMean, method = 'spearman')



res14 %>%
  ggplot(aes(y = novel_baseMean,x = normal_baseMean)) +
  geom_point(alpha = 0.2) +
  yscale("log10", .format = TRUE)+
  xscale("log10", .format = TRUE) +
  geom_abline(slope=1, intercept=0, color = 'red')
  
```

