---
title: "The GC and SS score of the non-AS"
author: "Anqi Dai" 
date: "5/23/2019" 
output: html_document
---  
     
```{r setup, include=FALSE}  
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(BSgenome)
library(BSgenome.Hsapiens.NCBI.GRCh38)
library(seqinr)    
```

## Load data for all the bg introns
 
```{r}
bg <- read_csv('../../15_true_and_false/data/unique_introns_for_expressive_genes.csv') %>% 
  filter(intronLen > 30) 

# the 75 quantile of sig IR events
sig75 <- 358
```

```{r}
# put the long and short difference in the table
bg <- bg %>% 
  mutate(category = if_else(intronLen < sig75, 'short','long') )
```

## Retrieve the seq for GC 

```{r}
genome <- BSgenome.Hsapiens.NCBI.GRCh38
seqlevelsStyle(genome) = "UCSC"
grange = makeGRangesFromDataFrame(bg, seqnames.field=c("chr"), start.field="start",
                                  end.field=c("end"), strand.field="strand")
seq = BSgenome::getSeq(genome, grange)
 
ret <- bg %>%
  mutate(Seq = as.character(seq))  %>%
  mutate(GC = stringr::str_count(Seq, "G") + stringr::str_count(Seq, "C")) %>%
  mutate(GC_per = GC / stringr::str_length(Seq) * 100)
```
 
## Retrieve the both ends seq for SS score
 
```{r}
# * Score 5' splice site: Each sequence must be 9 bases long. 3 bases in exon and 6 bases in intron
# * Score 3' splice site: Each sequence must be 23 bases long. 20 bases in the intron and 3 base in the exon
ret <-  bind_rows(
  ret %>%
    filter(strand == '+') %>%
    mutate(ss5_start = start - 3,
           ss5_end = start + 5,
           ss3_start = end - 19,
           ss3_end = end + 3 ),
  ret %>%
    filter(strand == '-') %>%
     mutate(ss5_start = end - 5,
           ss5_end = end + 3,
           ss3_start = start - 3,
           ss3_end = start + 19 ))

grange5 <-  makeGRangesFromDataFrame(ret, 
                                     seqnames.field="chr",  
                                     start.field="ss5_start", 
                                     end.field="ss5_end", 
                                     strand.field="strand")

grange3 <-  makeGRangesFromDataFrame(ret, 
                                     seqnames.field="chr",  
                                     start.field="ss3_start", 
                                     end.field="ss3_end", 
                                     strand.field="strand")

ret <- ret %>%
  mutate(ss5_seq = as.character(BSgenome::getSeq(genome, grange5)),
         ss3_seq = as.character(BSgenome::getSeq(genome, grange3)),
         id = paste(chr,paste(start, end, sep = '-'), sep = ':' ) )

# write out to fasta file
write.fasta(sequences = as.list(ret$ss5_seq), 
            names = ret$id,
            file.out =  '../data/bg.ss5.seq.fasta')
write.fasta(sequences = as.list(ret$ss3_seq), 
            names = ret$id,
            file.out =  '../data/bg.ss3.seq.fasta')
```
 
```{r}
# get the score from the website
maxent_res <- lapply(file.path('../output', list.files('../output/', pattern = 'bg')), function(fn) {
  res = read_delim(fn,comment = '>',delim = '\t',
                   col_names = F) %>%
    mutate(score = as.numeric(stringr::str_replace_all(string = X2, pattern = 'MAXENT: ', replacement = '')))
})

ret$ss5_score <- maxent_res[[2]]$score
ret$ss3_score <- maxent_res[[1]]$score
```
 
```{r}
# output the df 
ret <- ret %>%
  mutate(category = str_glue('Non-AS-{category}')) 

# some Seq seems empty not because the seqs are not retrieved but the seq is too long to fit into the space 

ret %>% 
  dplyr::select(chr, ss5_start:ss3_end, strand, locus:category, GC_per, ss5_score:ss3_score) %>% 
  write_csv('../data/03_non_AS_seq_info.csv')
  

```

